### DISCUSSION TITLE
Support Brahmic/Indic scripts

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#166] [Cross-project feature proposals]
# Support Brahmic/Indic scripts

![image](https://user-images.githubusercontent.com/53421382/208366339-ab1eb3c1-a599-4845-aaba-dca04da592de.png)
_image source https://en.wikipedia.org/wiki/Brahmic_scripts#/media/File:Different_scripts_of_different_languages_of_India.jpg_

Maplibre does not have support for all languages, in particular Brahmic/Indic scripts are not supported. This prevents adoption by anyone who needs a map in those languages, or use-cases where you want a map in the local language and want coverage of the world (e.g. Wikimedia and OSM maps).

This would also give MapLibre an advantage over Mapbox GL for those use-cases, as it still doesn't support them.

## Deliverables
- Support for Brahmic/Indic scripts in MapLibre GL JS
- Support for Brahmic/Indic scripts in MapLibre GL Native

This may necessitate major changes to how glyphs are handled.

Ref: https://github.com/mapbox/mapbox-gl-js/issues/4009

Originally posted by @pnorman

**Comment by answerquest** [comment]
What's the technical work that it'll take to make this happen?

**Reply by HarelM** [reply]
There's a need to update the existing code that draws letters to accommodate these languages that doesn't work like a typewriter (you need to go back and change things as you add more characters). This is very complicated as far as I know.
Mix that with the need to write words along lines and you get a hot mess of complexity... :-/
The writing is done in the GPU and you can't use the browser html support for these languages, unfortunately.
The RTL support can be seen here, it has some c code to support it:
https://github.com/mapbox/mapbox-gl-rtl-text
I personally have no clue how all this works together in depth. This is what I understood when I tried to get a high-level understanding of this issue.

**Reply by ramSeraph** [reply]
From what I have seen, the list is long and spread over multiple issue trackers and PRs across the mapbox-gl-js and node-fontnik repos. I think this should be a generic Complex-Text-Layout script support than just Brahmic/Indic script support.

A few options( like using the wasm port of harfbuzz, client side glyph SDF generation ) have opened up in the recent years which were not previously available. Mostly though, I think it is a matter of testing performance impact  and correctness of the various options available. 

Initially, maybe something similar to the RTL text plugin to minimise the performance impact on non-Complex-Text-Layout scripts and keep things backward compatible with flags. 

Also, may require server side changes, like not indexing the glyphs by char codes and performing unicode bidi processing and text breaking at the vector tile data level itself.

**Reply by ramSeraph** [reply]
> There's a need to update the existing code that draws letters to accommodate these languages that doesn't work like a typewriter (you need to go back and change things as you add more characters). This is very complicated as far as I know. Mix that with the need to write words along lines and you get a hot mess of complexity... :-/ The writing is done in the GPU and you can't use the browser html support for these languages, unfortunately. The RTL support can be seen here, it has some c code to support it: https://github.com/mapbox/mapbox-gl-rtl-text I personally have no clue how all this works together in depth. This is what I understood when I tried to get a high-level understanding of this issue.

Did not see this, yeah.. I agree, it is a mess :)

**Reply by planemad** [reply]
This ticket https://github.com/mapbox/mapbox-gl-native/issues/7774 by @chrisloer probably is a good technical overview of what needs to be done. Curious to know what related developments have happened in the last 5 years which might get us closer to the solution.

The [tangram-es](https://github.com/tangrams/tangram-es) renderer from Mapzen implemented support for Indic support in GL with the limitation of only horizontal label rendering. Demo: https://osm-in.github.io/indic-map/ . 

Even if we start with such a limited support in maplibre, it would still be a huge leap from the current state of broken rendering.

**Reply by HarelM** [reply]
It would be great if someone can take this on and maybe suggest a plugin like the RTL or a PR to maplibre-gl-js and maybe maplibre-native.
I'll be happy to review it from the maplibre-gl side, but there's a need for someone that knows the language and can test the corner cases and how they should look. I don't have the relevant knowledge and skills unfortunately.

**Reply by ramSeraph** [reply]
> This ticket https://github.com/mapbox/mapbox-gl-native/issues/7774 by @chrisloer probably is a good technical overview of what needs to be done. Curious to know what related developments have happened in the last 5 years which might get us closer to the solution.
> 
> The [tangram-es](https://github.com/tangrams/tangram-es) renderer from Mapzen implemented support for Indic support in GL with the limitation of only horizontal label rendering. Demo: https://osm-in.github.io/indic-map/ . 
> 
> Even if we start with such a limited support in maplibre, it would still be a huge leap from the current state of broken rendering.

In general, wasm support is more widespread in browsers now than it was 5 years back making it a more viable option.

Apart from that, the rust ecosystem for Unicode handling has improved a lot.. there is also rustybuzz for handling font shaping, there also seems to be some advancement in text break handling in the rust Unicode ecosystem which can pull per language dictionary data lazily. Given the ease of deploying rust code to the browser all of this can be used on the client side. 

Even coming from the c++ side, emscripten improvements might translate to smaller bundle sizes and faster code. 

There also seems to be browser APIs coming up for exposing font metrics and extracting SNFT data from local fonts. This is not widely spread though.

**Reply by ChrisLoer** [reply]
If it's helpful, there is still a `cloer_harfbuzz` branch on mapbox-gl-native that has my early prototype of using Harfbuzz (also, I think the issue #7774 directly links to the `coretext-hack` branch. Couple issues:

- The branch didn't get pulled over to MapLibre. All the work was done when Mapbox GL was fully open source, but obviously looking at code in the Mapbox GL repo is something to be careful with. I'm not sure what the policy is on how to handle that case.
- The prototype just used locally installed fonts, so didn't touch the complicated server-side requirements at all
- I was just starting at Mapbox when I worked on that branch, _and_ the code has gone through a lot of change since then: the branch probably doesn't give that much guidance to someone trying to implement it now, if they're familiar with the MapLibre symbol generation code.
- The performance cost of adding harfbuzz shaping (specifically tile load times) seemed big enough that we wouldn't want to have it enabled when it wasn't "necessary". This is also what led to the ICU-based RTL text support being split out into a plugin on the JS side, but that was a much easier case because (1) the cost was mainly just the increase in bundle size, and (2) the plugin could turn on/off with very little modification of the rest of the shaping code.

I believe Harfbuzz compiled to wasm with emscripten is now a pretty widely deployed solution.

**Reply by wipfli** [reply]
Thanks @ChrisLoer. We recently discussed scripts in the MapLibre Asia Call, see https://github.com/maplibre/maplibre/discussions/154

@alanchenboy would you be interested in giving a presentation about the solution that Grab has sometime in January 2023?

**Reply by alanchenboy** [reply]
OK, I will make a presentation.

**Comment by alanchenboy** [comment]
GrabMap adds TTF support in style and extends the GlyphID:


```
style:
    "glyphs": "https://api.stg-myteksi.com/map-tiles/v1/fonts/{fontstack}/{range}.pbf",
    "fonts": "https://api.stg-myteksi.com/map-tiles/v1/fonts/{fontstack}/{language}.ttf",
```

 ```
   enum GlyphIDType : short {
        FontPBF = 0x00,
        Khmer = 0x01,
        Myanmar = 0x02
    };

    union GlyphID {
        char32_t hash;
        struct {
            char16_t code;
            GlyphIDType type;
        } complex;
    };
```

So if the char string is a complex code point (Khmer or Myanmar), then the char string will be sent to harfbuzz, processed via the corresponding font file, and changed into the char index of the font file with position offsets. 
Then the font manager will render the char code base on the index of the font via the font file and FreeType and create the SDF format glyph image.
Use the char index with offset and the Glyph image created by Freetype in layout then the complex text could be
 drawn correctly.
<img width="650" alt="Screenshot 2022-12-23 at 11 22 06" src="https://user-images.githubusercontent.com/5135389/209264607-bf8fbcda-dd5a-4a8d-8bd4-b89d3a283363.png">

**Comment by wipfli** [comment]
Here is a cool writing systems overview map from wikipedia:

https://upload.wikimedia.org/wikipedia/commons/9/90/WritingSystemsOfTheWorld.svg

<img src="https://upload.wikimedia.org/wikipedia/commons/9/90/WritingSystemsOfTheWorld.svg" />

**Comment by wipfli** [comment]
@alanchenboy offers to give a presentation about how GrabMap renders the Khmer writing system with FreeType and HarfBuzz. We will hear @alanchenboy in the next MapLibre Eastern Call which is scheduled for Wednesday Feb 8th, 2023 8 AM to 9 AM CET. 

If you want to join the Khmer writing system presentation, please send me a message in slack. Join the slack `#maplibre` channel at https://slack.openstreetmap.us and ping me there...

**Comment by wipfli** [comment]
This project was selected to become a Bounty Direction. You find a tracking issue at https://github.com/maplibre/maplibre/issues/193
